### PR TITLE
docs(oep): proposal for local bind volumes (#3792)

### PR DESCRIPTION
**Pull request by todeb**

### CODE REVIEW

File: designs/helm-charts/local-bind-volumes.md
Comment (tiagolobocastro) [inline designs/helm-charts/local-bind-volumes.md:51]: I think you can leave Jiva and cStor out, as they're no longer OpenEBS engines (they're archived)

File: designs/helm-charts/local-bind-volumes.md
Comment (tiagolobocastro) [inline designs/helm-charts/local-bind-volumes.md:11]: We can set as provisional since it's in PR.
It can also be merged as many times as you wish whilst provisional

```suggestion
---
oep-number: Issue also needs to be created
title: OpenEBS Enhancement Proposal for Mayastor Local Feature
authors:
  - "@todeb"
owners:
  - "@todeb"
  - "@tiagolobocastro"
editor: TBD
creation-date: 2024-11-07
last-updated: 2024-11-07
status: provisional
---

# OpenEBS Enhancement Proposal for Mayastor Local Feature
```

File: designs/helm-charts/local-bind-volumes.md
Comment (tiagolobocastro) [inline designs/helm-charts/local-bind-volumes.md:46]: What if you set `local` on an existing volume, residing on node_x and existing pod residing on node_y?
What would be the expected behaviour? The pod consuming the volume would be moved only after existing pod is deleted? (assuming it's some kind of replica set and not a plain Pod without a controller)

File: designs/helm-charts/local-bind-volumes.md
Comment (tiagolobocastro) [inline designs/helm-charts/local-bind-volumes.md:103]: Do these rules help with `Immediate` binding mode? Also can they be modified after initial creation? (since local flag is intended to be modifiable via plugin)

**Review by tiagolobocastro — COMMENTED** [review]
Great proposal @todeb !
@openebs/org-maintainers please take the time to review this.
Other folks please also chip in.
@todeb you may merge this PR as required as we iterate, we don't need to agree all details until the final decision to take this in.

**Comment by tiagolobocastro** [comment]
@abhilashshetty04 shall we disable rust lint for PR's which don't modify rust code?

File: designs/helm-charts/local-bind-volumes.md
Comment (todeb) [reply designs/helm-charts/local-bind-volumes.md:46]: Do nothing, is up to user to rollout restart or delete pod and kubernetes scheduler will reshedule it then with affinity to local volume.

File: designs/helm-charts/local-bind-volumes.md
Comment (todeb) [reply designs/helm-charts/local-bind-volumes.md:104]: local flag is intended to be set in storageclass, and also be modified via plugin.
If volume is created with Immediate and local, then volume can be created on random node and set with proper affinity. K8S scheduler will then schedule pod on specific node.
```
Node Affinity:
  Required Terms:
    Term 0:        openebs.io/csi-node in [mayastor]
```
```
Node Affinity:
  Required Terms:
    Term 0:        openebs.io/nodename=node1
```

**Review by todeb — COMMENTED** [review]
did some clarifications in commit, let me know if sth more is missing.

File: designs/helm-charts/local-bind-volumes.md
Comment (Abhinandan-Purkait) [reply designs/helm-charts/local-bind-volumes.md:46]: How do we set PV affinity to existing volumes? The affinity is immutable in the PV. Also for a local:true volume, even if we set local to false, once PV affinity is set the application cannot be rescheduled to a new node. So how does the unset flow work?

Comment (todeb) [reply designs/helm-charts/local-bind-volumes.md:46]: right that might be the issue, possible options I see:
- get rid of the unset, once volume is created cannot be changed.
- delete pv definition retaining data and recreate the new pv
- make a volume replica with openebs.io/csi-node in [mayastor] affinity, and delete original volume

**Comment by abhilashshetty04** [comment]
> @abhilashshetty04 shall we disable rust lint for PR's which don't modify rust code?

I think we should.

File: designs/helm-charts/local-bind-volumes.md
Comment (tiagolobocastro) [reply designs/helm-charts/local-bind-volumes.md:108]: I see, it kinda reverses the initial step, so basically:
`WaitForFirsConsumer`: brings storage to node where pod is first scheduled on
`Immediate`: brings pod to the node where mayastor chooses to place the replica

Comment (todeb) [reply designs/helm-charts/local-bind-volumes.md:108]: yes, and what is the confusion or reverse?

File: designs/helm-charts/local-bind-volumes.md
Comment (todeb) [reply designs/helm-charts/local-bind-volumes.md:50]: unset is not mandatory to be implemented for this, but could be nice to have add on.
3rd point right, it doesn't make any sense as the pv definition has the nodeaffinity and not the mayastor itself, so making more replicas of volume won't change anything in that case.

File: designs/helm-charts/local-bind-volumes.md
Comment (tiagolobocastro) [reply designs/helm-charts/local-bind-volumes.md:108]: It's clear for me now, just wanted to highlight this difference, which I think is important.

**Comment by Abhinandan-Purkait** [comment]
@todeb You would need to sign your commits for DCO. You can refer the below for steps.
https://github.com/openebs/openebs/pull/3801/checks?check_run_id=32713394687

File: designs/helm-charts/local-bind-volumes.md
Comment (tiagolobocastro) [reply designs/helm-charts/local-bind-volumes.md:50]: > 3rd point right, it doesn't make any sense 

ok...
Not sure where "making more replicas" is coming from?
My point here is that with a custom scheduler you don't need to use the CSI topology feature, at all. 
At the cost of having to maintain a custom scheduler, which we might want to avoid..

Comment (tiagolobocastro) [reply designs/helm-charts/local-bind-volumes.md:50]: oh I see, your 3rd point, right yes I think replicas won't help here.
Anyways my suggestion is to keep this simple, and remove the ability to modify the local for now, that perhaps can be tackled on separate oep?

**Review by tiagolobocastro — APPROVED** [review]
IMHO we should remove local modification to keep this simpler, or add it as a milestone before we make a case to bring this into the roadmap

**Comment by tiagolobocastro** [comment]
I think we can merge first cut of this, @todeb would you be able to move this file from helm-charts\ into replicated-pv\mayastor ?

**Comment by todeb** [comment]
> I think we can merge first cut of this, @todeb would you be able to move this file from helm-charts\ into replicated-pv\mayastor ?

moved, I'm not authorized, so you can merge this.
You’re not [authorized](https://docs.github.com/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branches) to merge this pull request.

**Comment by avishnu** [comment]
@todeb would you be interested in taking this to the next level? This means, requesting the maintainers for consideration of the proposed OEP. Also, if you have further thoughts on this, please feel free to modify the proposal.

**Comment by fara-tode** [comment]
@avishnu I don't really know how to handle this, I'm not developer.
But I'm actually seeing another problem were bind is not used. 

When doing restore by velero, mayastor will create volume on "random" node, and velero will create pod handling restore process on "random" node too. Usually these "random" nodes are different, and this slows significally restore time for volume.

**Comment by todeb** [comment]
@avishnu I'm not a developer, so I'm not entirely sure how to proceed with requesting consideration from the maintainers.

I’ve noticed another issue related to how bind is not being used. During a restore with Velero, Mayastor often creates the volume on a "random" node, while Velero schedules the restore pod on a different "random" node. When these nodes don’t match, restore performance is significantly degraded because IOPS are much lower when accessing the volume remotely.

**Pull request merged**
